Prevents read-only transactions from trigger tx events

tinwelint opened the pull request:
This was a behaviour change introduced in 2.2 by mistake.

Additionally (in a second commit):
Prevents token mini-transactions from triggering tx events

since that causes confusion and the TransactionData cannot access
information about created tokens anyway.

tinwelint committed:
Prevents read-only transactions from triggering tx events

This was a behaviour change introduced in 2.2 by mistake

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java:
```diff
@@ -458,10 +460,20 @@ private void commit() throws TransactionFailureException
-            // Trigger transaction "before" hooks
-            if ( (hooksState = hooks.beforeCommit( txState, this, storeLayer )) != null && hooksState.failed() )
+            // Trigger transaction "before" hooks.
+            if ( hasChanges() )
```
Will that mean that in case of read only transaction your afterCommit or afterRollback hook will never be invoked?

tinwelint replied:
Exactly, since there's not really anything that was committed, right?

MishaDemianenko replied:
From a perspective of affected data fully agree, in the same time if you have a hook that lets say track last successful transaction commit time you will not be able to get that and in the same time your rollback hook will not gonna be invoked also. Which looks confusing. Do you agree?

tinwelint replied:
I don't understand your concern. Could you elaborate?

MishaDemianenko replied:
`TransactionEventHandler` describes methods that can be used to listen certain type of _transaction_ events, and for example transaction can be committed irrespectively to operation types that it just performed. 
I also can see that in java doc it says that read only transactions will not be notified and previous behaviour was a bug according to that description, but still that sounds strange to me.
I still can imagine use cases when people would like to be notified when transaction was committed or rollbacked disregard disregard it was read only.

tinwelint replied:
Although what would the purpose of getting such notification be? There would be no data to try and figure out what the transaction was about. The only useful thing would be to look at the stack trace to see where the transaction came from... doesn't strike me as useful.

MishaDemianenko replied:
I can imagine that it can be useful for monitoring purposes - when you track last commit/rollback transaction time or count number of successful or failed transactions

tinwelint replied:
Sure, but I don't see that as part of a transaction event handler. See [TransactionMonitor](https://github.com/neo4j/neo4j/blob/2.3/community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/TransactionMonitor.java)

tinwelint committed:
Prevents token mini-transactions from triggering tx events

since that causes confusion and the TransactionData cannot access
information about created tokens anyway.
